### DISCUSSION TITLE
github webhook push Delivery Fail.

### DISCUSSION THREAD

**Discussion by oliverpark999** [argoproj/argo-cd#11851] [Q&A]
I couldn't wait for 3 minutes to be synced, so I setting the github webhook.
But I'm repeating 'success-fail' like a screenshot.
Sometimes I keep saying "success" and sometimes I fail.
What the hell is wrong with you?

**Error Code**
```
400
Body
Webhook processing failed: HMAC verification failed
```
<img width="463" alt="githubwebhook" src="https://user-images.githubusercontent.com/47516457/209897890-111929b0-6bc6-4b58-a854-16e6ad6b5f76.png">

**Comment by batistein** [comment]
I'm encountering the same issue. Any updates?

**Reply by sjwoodr** [reply]
I'm seeing the same behavior here.

**Reply by marcportabellaclotet-mt** [reply]
Same here, but randomly

**Reply by sjwoodr** [reply]
We eliminated this problem entirely it seems.  The webhook secret was base64 encoded via `echo xyz | base64` and after we changed it to `echo -n xyz | base64` the problem went away and hasn't resurfaced.  It doesn't explain why including the trailing newline in the secret works *sometimes* but not all the time, but that seems to have fixed it for us (not including the newline at all).

**Reply by reefland** [reply]
Same issue...  already using `echo -n xyz | base64`.  Worked great yesterday when initially put in place for testing, but 100% failure rate now with `Webhook processing failed: HMAC verification failed`.

![image](https://github.com/argoproj/argo-cd/assets/16230914/b90c2ad7-84cd-469c-ab1d-f2d4e3a936d3)

**Comment by VictorGaiva** [comment]
For anyone with this issue, I found out that you can't use a Webhook that have some special characters, probably `+`, `/` or `=`. Since I was generating them using `openssl rand -base64 32`, it could happen to have these characters. 

A workaround is to keep generating them until some is generated without it, them adding it to `argocd-secret` `stringData` field, and the same value to Github webhook configuration.

I had the same issue when I was configuring `Postgresql`'s password using `crunchydata/pgo`

**Reply by aydinseven7** [reply]
I can't believe I spent an hour trying to troubleshoot the issue...I only used `$` and `^` though, removing them did the trick.

**Reply by chanakya-svt** [reply]
I am still facing the issue. Following are the steps I have taken and still same issue.
- Setup secret in GitHub webhook. For example: `76af18c59313` 
- updated the `argocd-secret` in `argocd` namespace and added `stringData -> webhook.github.secret` with the above value(`76af18c59313`)
- restarted the `argocd-server` pod

After all these steps, the GitHub webhook push still fails.

If I remove the secret on both sides then they work as expected.Adding the secret is causing `Webhook processing failed: HMAC verification failed` all the time.

**Reply by VictorGaiva** [reply]
I remember going through an issue with argocd ignoring the password I setup on the secret, but I can't recall exactly had happened.

**Reply by bogd** [reply]
I had a similar issue here (`HMAC verification failed`, even after removing all special characters from secret). It worked fine without a secret, but as soon as I added one I would start getting the error.

In my case, it was solved by an upgrade of the ArgoCD server (I was running v2.7, and upgraded to 2.11). I guess some algorithm was changed at some point?

**Reply by thiagowfx** [reply]
We are experiencing the same issue. Our Argo setup is done via its official helm chart, therefore it's not possible to use `echo -n ... | base64`  as suggested by an earlier message. Even by removing special characters, it still happens intermittently. We are using v2.11+.

**Comment by AndreiPetrusMihai** [comment]
I'm leaving this here for people that encounter this problem after changing/first setting the webhook secret. Make sure to restart the server pods so that they use the newly set value.
